src/stream/src/executor/managed_state/aggregation/extreme.rs

```diff
@@ -297,15 +316,40 @@ where
             // To future developers: please make **SURE** you have taken `EXTREME_TYPE` into
             // account. EXTREME_MIN and EXTREME_MAX will significantly impact the
             // following logic.
-            let all_data = self.keyspace.scan(self.top_n_count, epoch).await?;
-
-            for (raw_key, mut raw_value) in all_data {
-                // let mut deserializer = value_encoding::Deserializer::new(raw_value);
-                let value = deserialize_cell(&mut raw_value, &self.data_type)
-                    .map_err(StreamExecutorError::serde_error)?;
-                let key = value.clone().map(|x| x.try_into().unwrap());
-                let pks = self.serializer.get_pk(&raw_key[..])?;
-                self.top_n.insert((key, pks), value);
+
+            let all_data_iter = state_table
+                .iter_with_pk_prefix(
+                    self.group_key.as_ref().unwrap_or(&Row::default()),
+                    OrderedRowSerializer::new(vec![
+                        if EXTREME_TYPE == variants::EXTREME_MAX {
+                            OrderType::Descending
+                        } else {
+                            OrderType::Ascending
+                        };
+                        self.group_key
+                            .as_ref()
+                            .map_or(0, |key| key.size())
+                    ]),
+                    epoch,
+                )
+                .await?;
+            pin_mut!(all_data_iter);
+
+            for _ in 0..self.top_n_count.unwrap_or(usize::MAX) {
+                if let Some(inner) = all_data_iter.next().await {
+                    let row = inner.unwrap().into_owned();
+                    let value = row[row.0.len() - 1].clone();
+                    let mut key = ExtremePk::with_capacity(1);
+                    let group_key_len = self.group_key.as_ref().map_or(0, |row| row.size());
```

Comment:
This part still looks a bit ungly. I will rewrite it more rusty. But the main idea should be same